src/main/java/me/zeroeightsix/fiber/builder/ConfigValueBuilder.java

```diff
@@ -118,69 +123,82 @@ public ConfigValueBuilder(@Nonnull String name, @Nonnull Class<T> type) {
+    public ConfigValueBuilder<T> withFinality() {
         this.isFinal = true;
         return this;
     }
 
     /**
      * Sets the finality.
      *
-     * <p> If {@code true}, the produced setting can not be changed. It will be initialised with its default value, if there is one. Afterwards, it can not be changed again.
+     * <p> If {@code true}, the produced setting can not be changed.
+     * It will be initialised with its default value, if there is one. Afterwards, it can not be changed again;
+     * {@link ConfigValue#setValue(Object)} will always return {@code false}.
      *
-     * @param isFinal the finality
+     * @param isFinal whether or not the value can be changed after building
      * @return {@code this} builder
      */
-    public ConfigValueBuilder<T> setFinal(boolean isFinal) {
+    public ConfigValueBuilder<T> withFinality(boolean isFinal) {
         this.isFinal = isFinal;
         return this;
     }
 
-    /**
-     * Sets the node that the {@code ConfigValue} will be registered to.
-     *
-     * @param node The node the {@link ConfigValue} will be registered to.
-     * @return The builder
-     */
-    public ConfigValueBuilder<T> withParent(Node node) {
-        parentNode = node;
-        return this;
-    }
-
     /**
      * Creates a constraint builder for this {@code ConfigValueBuilder}.
      *
      * @return the created builder
      * @see ConstraintsBuilder
      */
-    public ConstraintsBuilder<? extends ConfigValueBuilder<T>, T> constraints() {
+    public ConstraintsBuilder<T> beginConstraints() {
         return new ConstraintsBuilder<>(this, constraintList, type);
     }
 
     /**
      * Builds the {@code ConfigValue}.
      *
-     * <p> If a parent was specified using {@link #withParent}, the {@code ConfigValue} will also be registered to its parent node.
+     * <p> If a parent was specified in the constructor, the {@code ConfigValue} will also be registered to its parent node.
+     *
+     * <p> <strong>This method should not be called multiple times if the default value is intended to be mutated!</strong>
+     * Multiple calls will result in duplicated references to the default value.
      *
      * @return the {@code ConfigValue}
      */
     public ConfigValue<T> build() {
-        ConfigValue<T> built = new ConfigValue<>(name, comment, defaultValue, defaultValue, consumer, constraintList, type, isFinal);
+        if (defaultValue != null) {
+            for (Constraint<? super T> constraint : constraintList) {
+                if (!constraint.test(defaultValue)) {
+                    throw new RuntimeFiberException("Default value '" + defaultValue + "' does not satisfy constraints");
+                }
+            }
+        }
+        List<Constraint<? super T>> constraints = new ArrayList<>(this.constraintList);
+        if (isFinal) {
+            constraints.add(0, FinalConstraint.instance());  // index 0 to avoid uselessly checking everything each time
+        }
```

Comment:
Because the 'root constraint' defaults to `AND`, you can just remove the other constraints. `false && someBoolean` will always be false.

Reply:
Maybe someone would still like to inspect the other constraints ? Like with a `version` field, you may want finality and regex, and a third party tool could want to pick up that.

Reply:
Also if we move towards mutability, we have to consider the possibility of something removing the `final` constraint after build.

Reply:
That's reasonable - you can just leave it as is.